### ISSUE TITLE
`var_list` will cause untrainable variables to be trained

### ISSUE THREAD

**Issue by machinaut** [tensorflow/tensorflow#23639]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): below
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  Mac 10.13
- TensorFlow installed from (source or binary): `pip`
- TensorFlow version (use command below): `v1.9.0-0-g25c197e023 1.9.0`
- Python version: `3.6.5`

You can collect some of this information using our environment capture [script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

```v1.9.0-0-g25c197e023 1.9.0```

**Describe the current behavior**
Both trainable and untrainable variables are updated by the optimizer

**Describe the expected behavior**
Expected: trainable variables are updated by optimizer, but untrainable variables are not.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

```
import tensorflow as tf

x = tf.Variable([1.0], trainable=False)
y = tf.Variable([1.0], trainable=True)
z = x + y

train_op = tf.train.GradientDescentOptimizer(0.1).minimize(z, var_list=[x, y])
sess = tf.Session()
sess.run(tf.global_variables_initializer())
sess.run(train_op)

print(sess.run(x))  # [0.9]  Unexpected
print(sess.run(y))  # [0.9]  Expected
```

**Comment by unrealwill** [comment]
@machinaut I think this is intended behavior and not a bug. Trainable is just a flag to help manage variables collections.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @wt-huang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by gadagashwini** [comment]
Hi @machinaut,
When building a machine learning model it is often convenient to distinguish between variables holding trainable model parameters and other variables such as a step variable used to count training steps. To make this easier, the variable constructor supports a trainable=<bool> parameter. [tf.GradientTape](https://www.tensorflow.org/api_docs/python/tf/GradientTape) watches trainable variables by default:

```
with tf.GradientTape(persistent=True) as tape:
  trainable = tf.Variable(1.)
  non_trainable = tf.Variable(2., trainable=False)
  x1 = trainable * 2.
  x2 = non_trainable * 3.
tape.gradient(x1, trainable)
```
`<tf.Tensor: shape=(), dtype=float32, numpy=2.0>`

`assert tape.gradient(x2, non_trainable)`  is `None`  # Unwatched

Thank you!
